tests/test_helper.tcl

```diff
@@ -585,21 +609,40 @@ proc signal_idle_client fd {
     } else {
         lappend ::idle_clients $fd
         set ::active_clients_task($fd) "SLEEPING, no more units to assign"
-        if {[llength $::active_clients] == 0} {
+        if {[llength $::active_clients] == 0 && $::loop <= 1} {
             the_end
         }
     }
+
+    foreach idle_fd $clients_to_wake {
+        signal_idle_client $idle_fd
+    }
 }
 
 # The the_end function gets called when all the test units were already
 # executed, so the test finished.
+proc print_test_summary {} {
+    set out "\nTest Summary: [colorstr bold-green $::ok_tests_count] passed, [colorstr bold-red $::err_tests_count] failed"
+    if {$::loop > 0} {
+        append out "\nTest Loops Summary: [colorstr bold-green $::loop_passes] passed, [colorstr bold-red $::loop_failures] failed"
+    }
+    puts $out
+}
+
 proc the_end {} {
+    if {$::loop_failed_current} {
+        incr ::loop_failures
+    } else {
+        incr ::loop_passes
+    }
+    set ::loop_failed_current 0
```

Comment:
Could we move it up the stack ? `the_end` can be used just kept for logging the summary.